File: src/main/java/nz/co/searchwellington/controllers/ContentController.java
package nz.co.searchwellington.controllers;

import java.io.IOException;

import javax.servlet.http.HttpServletRequest;
import javax.servlet.http.HttpServletResponse;

import nz.co.searchwellington.annotations.Timed;
import nz.co.searchwellington.controllers.models.ContentModelBuilderService;

import org.apache.log4j.Logger;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Controller;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.servlet.ModelAndView;

import com.sun.syndication.io.FeedException;

@Controller
public class ContentController {

    private static final Logger log = Logger.getLogger(ContentController.class);

    private ContentModelBuilderService contentModelBuilder;
    private UrlStack urlStack;

    public ContentController() {
    }

    @Autowired
    public ContentController(ContentModelBuilderService contentModelBuilder, UrlStack urlStack) {
        this.contentModelBuilder = contentModelBuilder;
        this.urlStack = urlStack;
    }

    @RequestMapping(value = {"/", "/*", "/search", "/archive/*/*", "/*/comment", "/*/geotagged", "/feed/*", "/feeds/inbox", "/tags", "/tags/json", "/*/json", "/*/rss", "/*/*/*/*/*"})
    @Timed(timingNotes = "")
    public ModelAndView normal(HttpServletRequest request, HttpServletResponse response) throws IllegalArgumentException, FeedException, IOException {
        final ModelAndView mv = contentModelBuilder.populateContentModel(request);
        if (mv != null) {
            if (isHtmlView(mv)) {
                urlStack.setUrlStack(request);
            }
            return mv;
        }

        log.warn("Model was null; returning 404");
        response.setStatus(HttpServletResponse.SC_NOT_FOUND);
        return null;
    }

    private boolean isHtmlView(ModelAndView mv) {
        return mv.getViewName() != null;
    }

}

File: src/main/scala/nz/co/searchwellington/controllers/ContentController.scala
package nz.co.searchwellington.controllers

import io.opentelemetry.api.trace.Span
import nz.co.searchwellington.ReasonableWaits
import nz.co.searchwellington.controllers.models.ContentModelBuilderServiceFactory
import nz.co.searchwellington.filters.RequestFilter
import org.apache.commons.logging.LogFactory
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.core.annotation.Order
import org.springframework.http.HttpStatus
import org.springframework.stereotype.Controller
import org.springframework.web.bind.annotation.GetMapping
import org.springframework.web.server.ResponseStatusException
import org.springframework.web.servlet.ModelAndView

import javax.servlet.http.{HttpServletRequest, HttpServletResponse}
import scala.concurrent.{Await, ExecutionContext}

@Order(5)
@Controller
class ContentController @Autowired()(contentModelBuilderServiceFactory: ContentModelBuilderServiceFactory,
                                     urlStack: UrlStack,
                                     requestFilter: RequestFilter,
                                     loggedInUserFilter: LoggedInUserFilter) extends ReasonableWaits {

  private val log = LogFactory.getLog(classOf[ContentController])

  private val contentModelBuilderService = contentModelBuilderServiceFactory.makeContentModelBuilderService()

  // Ant-style path patterns
  @GetMapping(path = Array("/", "/*", "/search", "/archive/*", "/profiles/**", "/*/geotagged", "/feed/*",
    "/feed/*/json", "/feeds/inbox", "/profiles", "/publishers", "/publishers/json", "/tags", "/tags/json", "/*/json", "/*/rss",
    "/*/geotagged/rss", "/*/geotagged/json",
    "/newsitem/*",
    "/{\\w+}/{year:\\d+}-{month:\\w+}"
  ))
  def normal(request: HttpServletRequest, response: HttpServletResponse): ModelAndView = {
    import scala.concurrent.ExecutionContext.Implicits.global
    implicit val currentSpan: Span = Span.current()
    buildAndRender(request, response)
  }

  private def buildAndRender(request: HttpServletRequest, response: HttpServletResponse)(implicit ec: ExecutionContext, currentSpan: Span): ModelAndView = {
    val loggedInUser = loggedInUserFilter.getLoggedInUser
    val eventualMaybeView = requestFilter.loadAttributesOntoRequest(request).flatMap { attributes =>
      attributes.toSeq.foreach( a =>
      request.setAttribute(a._1, a._2)
      )
      contentModelBuilderService.buildModelAndView(request, response, loggedInUser)
    }

    try {
      Await.result(eventualMaybeView, TenSeconds).fold {
        log.warn("Model was null; returning 404")
        throw new ResponseStatusException(HttpStatus.NOT_FOUND, "Not found")

      } { mv =>
        if (isHtmlView(mv)) {
          urlStack.setUrlStack(request)
        }
        mv
      }
    } catch {
      case r: ResponseStatusException =>
        throw r
      case e: Exception =>
        log.error("Error building content", e)
        throw e
    }
  }

  private def isHtmlView(mv: ModelAndView): Boolean = mv.getViewName != null

}
